apps/hash-frontend/src/pages/new/types/entity-type/select-generated-property-types.tsx

```diff
@@ -161,27 +170,35 @@ export const SelectGeneratedPropertyTypes: FunctionComponent<
 
   const handlePropertyTypeCheckboxChange = ({
     target,
-  }: React.ChangeEvent<HTMLInputElement>) => {
-    setSelectedPropertyTypeTitles((prev) =>
-      target.checked
-        ? [...prev, target.name]
-        : prev.filter((title) => title !== target.name),
-    );
-  };
+  }: ChangeEvent<HTMLInputElement>) => {
+    setGeneratedPropertyTypeDefinitions((prev) => {
+      const existingPropertyTypeDefinitionIndex =
+        prev?.findIndex(({ definition: { title } }) => title === target.name) ??
+        -1;
 
-  useEffect(() => {
-    if (generatedPropertyTypeDefinitions) {
-      setSelectedPropertyDefinitions(
-        generatedPropertyTypeDefinitions.filter(({ title }) =>
-          selectedPropertyTypeTitles.includes(title),
-        ),
-      );
-    }
-  }, [
-    selectedPropertyTypeTitles,
-    generatedPropertyTypeDefinitions,
-    setSelectedPropertyDefinitions,
-  ]);
+      const updatedSelectedPropertyTypes =
+        prev && existingPropertyTypeDefinitionIndex >= 0
+          ? [
+              ...prev.slice(0, existingPropertyTypeDefinitionIndex),
+              {
+                ...prev[existingPropertyTypeDefinitionIndex]!,
+                selected: !prev[existingPropertyTypeDefinitionIndex]!.selected,
+              },
+              ...prev.slice(existingPropertyTypeDefinitionIndex + 1),
+            ]
+          : prev;
+
+      if (generatedPropertyTypeDefinitions) {
+        onSelectedPropertiesChange(
+          updatedSelectedPropertyTypes
+            ?.filter(({ selected }) => selected)
+            .map(({ definition }) => definition) ?? [],
```

Comment:
This should be triggered outside of the set state function. You don't need to use the function syntax for set state as  `prev` will be identical to `generatedPropertyTypeDefinitions` in this context.

Reply:
Sure, have refactored in https://github.com/hashintel/hash/pull/2486/commits/a12868c77e15913f5c4076dc0013247c5820f939